src/core/BigDecimal.ts

```diff
@@ -0,0 +1,364 @@
+        let exp = 0; // exponent
+        let c: string; // current character
+        // integer significand array & idx is the index to it.
+        let idx = 0;
+        const coeff = [];
+        for (; len > 0; offset++, len--) {
+            c = value[offset];
+            // have digit
+            if (c >= '0' && c <= '9') {
+                // First compact case, we need not to preserve the character
+                // and we can just compute the value in place.
+                if (c === '0') {
+                    if (precision === 0) {
+                        coeff[idx] = c;
+                        precision = 1;
+                    } else if (idx !== 0) {
+                        coeff[idx++] = c;
+                        precision++;
+                    }  // else c must be a redundant leading zero
+                } else {
+                    if (precision !== 1 || idx !== 0) {
+                        precision++; // prec unchanged if preceded by 0s
+                    }
+                    coeff[idx++] = c;
+                }
+                if (dot) {
+                    scale++;
+                }
+                continue;
+            }
+            // have dot
+            if (c === '.') {
+                if (dot) { // two dots
+                    throw new RangeError('String contains more than one decimal point.');
+                }
+                dot = true;
+                continue;
+            }
+            // exponent expected
+            if ((c !== 'e') && (c !== 'E')) {
+                throw new RangeError('String is missing "e" notation exponential mark.');
+            }
+            exp = BigDecimal.parseExp(value, offset, len);
+            if (!Number.isSafeInteger(exp)) {
+                throw new RangeError('Exponent overflow');
+            }
+            break;
+        }
+        // here when no characters left
+        if (precision === 0) {
+            throw new RangeError('No digits found.');
+        }
+        // Adjust scale if exp is not zero.
+        if (exp !== 0) {
+            scale = BigDecimal.adjustScale(scale, exp);
+        }
+        const stringValue = coeff.join('');
+        if (isneg) {
+            unscaledValue = BigInt('-' + stringValue);
+        } else {
+            unscaledValue = BigInt(stringValue);
+        }
+        return new BigDecimal(unscaledValue, scale);
+    }
+
+    /**
+     * Adjust scales according to exp number.
+     * @param scale scale
+     * @param exp exponent number
+     * @throws RangeError if scale is out of range
+     */
+    private static adjustScale(scale: number, exp: number): number {
+        const adjustedScale = scale - exp;
+        if (!Number.isSafeInteger(adjustedScale)) {
+            throw new RangeError('Scale is out of range.');
+        }
+        scale = adjustedScale;
+        return scale;
+    }
+
+    /**
+     * Returns absolute value of a BigInt
+     * @param val
+     * @internal
```

Comment:
Should we put `internal` to private functions? If so, `adjustScale` is missing one. If it is not necessary, I am OK with leaving them as it is or removing them

Reply:
putting `internal` removes it from emitted type definition files, I will add to adjustScale as well. Thanks for the heads up